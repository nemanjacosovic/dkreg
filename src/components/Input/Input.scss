@import "src/constants/colors.module";

.dkreg-form-block {
  position: relative;
  padding-bottom: 15px;

  &__label {
    color: #666666;
    display: block;
    font-size: 12px;
    line-height: 18px;
    padding: 0 0 5px 0;
  }

  &__input-wrapper {
    display: block;
    position: relative;

    &.no-label {
      padding-top: 23px;
    }
  }

  &__input {
    z-index: 50;
    width: 100%;
    font-size: 14px;
    padding: 8px 16px;
    position: relative;
    border-radius: 3px;
    letter-spacing: 0.04em;
    background: $color-beta;
    transition: all ease-out 250ms;
    border: 1px solid rgba(209, 213, 219, 0.25);
  }

  &__icon {
    top: 8px;
    right: 10px;
    z-index: 60;
    overflow: auto;
    cursor: pointer;
    position: absolute;

    svg {
      fill: $color-gray-40;

      &:hover {
        opacity: 0.8;
      }
    }
  }

  &__helper {
    &--hint {
      padding: 0 8px;
      font-size: 11px;
      color: $color-gray-40;

      strong {
        font-weight: 700;
      }
    }

    &--error {
      font-size: 12px;
      line-height: 22px;
      color: $color-alert;

      svg {
        fill: $color-alert;
        display: inline-block;
        vertical-align: text-top;
      }
    }
  }
}
