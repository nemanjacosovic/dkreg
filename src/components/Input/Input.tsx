import React, { PureComponent } from 'react';
import InputConstants from '../../constants/CommonConstants';
import {IInputProps} from "../../interfaces/InputInterfaces";
import IconExclamationCircle from "../../assets/Icons/IconExclamationCircle";
import IconEye from "../../assets/Icons/IconEye";
import IconEyeClosed from "../../assets/Icons/IconEyeClosed";

import './Input.scss';

interface State {
    value: string;
    isShowPassword: boolean;
}

class Input extends PureComponent<IInputProps<HTMLInputElement>, State> {
    constructor(props: IInputProps<HTMLInputElement>) {
        super(props);

        this.state = {
            value: '',
            isShowPassword: false
        };
    };

    public onShowPassword = () => this.setState({isShowPassword: !this.state.isShowPassword});

    _renderLabel = () => {
        const {inputId, labelText} = this.props;

        if (!labelText) {
            return null;
        }

        return (
            <label htmlFor={inputId} className="dkreg-form-block__label">{labelText}</label>
        );
    };

    _renderInput = () => {
        const {isShowPassword} = this.state;
        const {inputType, placeholderText, inputName, inputRef, inputId, inputClassName} = this.props;

        const parseClassName = `dkreg-form-block__input${inputClassName ? ' ' + inputClassName : ''}`;

        return (
            <input
                type={isShowPassword ? InputConstants.TEXT : inputType}
                placeholder={placeholderText}
                name={inputName}
                ref={inputRef}
                id={inputId}
                className={parseClassName}
            />
        );
    };

    _renderViewEye = () => {
        const {isShowPassword} = this.state;

        if (!isShowPassword) {
            return <IconEyeClosed svgHeight={InputConstants.ICON_SIZE}/>
        }

        return <IconEye svgHeight={InputConstants.ICON_SIZE}/>
    };

    _renderInputComponent = () => {
        const {inputType, labelText} = this.props;

        const parseClassName = `dkreg-form-block__input-wrapper${!labelText ? ' no-label' : ''}`;

        if (inputType === InputConstants.PASSWORD) {
            return (
                <span className={parseClassName}>
                    {this._renderInput()}
                    <span className="dkreg-form-block__icon" onClick={this.onShowPassword}>
                        {this._renderViewEye()}
                    </span>
                </span>
            );
        }

        return (
            <span className={parseClassName}>{this._renderInput()}</span>
        );
    };

    _renderInputHint = () => {
        const { inputHint } = this.props;
        console.log('inputHint', inputHint);

        if (!inputHint) {
            return null;
        }
        return (
            <span className="dkreg-form-block__helper--hint">Du kan bruge bogstaver, tal og punktum.</span>
        )
    };

    _renderInputError = () => {
        const { inputError } = this.props;

        if (!inputError) {
            return null;
        }

        return (
            <span className="dkreg-form-block__helper--error">
                <IconExclamationCircle svgHeight="18"/>
                {inputError}
            </span>
        );
    };

    _renderHintOrError = () => {
        const isError = false;
        return (
            <div className="dkreg-form-block__helper">
                {isError ? this._renderInputError() : this._renderInputHint()}
            </div>
        );
    };

    render() {
        const {className} = this.props;

        return (
            <div className={`dkreg-form-block${className ? ' dkreg-form-block--' + className : ''}`}>
                {this._renderLabel()}
                {this._renderInputComponent()}
                {this._renderHintOrError()}
            </div>
        );
    }
}

export default Input;
