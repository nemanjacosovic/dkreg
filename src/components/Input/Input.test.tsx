import React from 'react';
import { shallow } from 'enzyme';

import InputConstants, { LanguageConstant } from "../../constants/CommonConstants";

import Input from './Input';

const testProps = {
    className: 'input-name-first',
    inputId: 'nameFirst',
    inputName: 'nameFirst',
    inputError: {
        message: LanguageConstant.NAME_FIRST_ERROR
    },
    inputType: InputConstants.TEXT,
    labelText: LanguageConstant.NAME_FIRST,
    placeholderText: LanguageConstant.NAME_FIRST_PLACEHOLDER
}

describe('<Input/> component', () => {
    let component = shallow(<Input {...testProps}/>);
    expect(component).toMatchSnapshot();

    it('<Input/> should exist', () => {
        expect(component.instance()).toBeDefined();
        expect(component).toHaveLength(1);
    });

    it('<Input/> should have default css classes', () => {
        expect(component.find('div').length).toBe(2);
        expect(component.find('input').length).toBe(1);
        expect(component.find('.dkreg-form-block').length).toBe(1);
        expect(component.find('.dkreg-form-block__input-wrapper').length).toBe(1);
    });

    it('<Input/> should render hint or error', () => {
        expect(component.find('.dkreg-form-block__helper').length).toBe(1);
        expect(component.find('.dkreg-form-block__helper--error').length).toBe(1);
        expect(component.find('.dkreg-form-block__helper--hint').length).toBe(0);
    });


    describe('rendering <Input/> as password', () => {
        beforeEach(() => {
            const testPropsNew = { ...testProps, inputType: InputConstants.PASSWORD, labelText: LanguageConstant.PASSWORD };
            component = shallow(<Input {...testPropsNew}/>);
        });

        it('<Input/> should exist', () => {
            expect(component.instance()).toBeDefined();
            expect(component).toHaveLength(1);
        });

        it('<Input/> should have css classes', () => {
            expect(component.find('IconEyeClosed').length).toBe(1);
            expect(component.find('IconExclamationCircle').length).toBe(1);
            expect(component.find('span').length).toBe(3);
            expect(component.find('.dkreg-form-block__icon').length).toBe(1);
        });

        it('<Input/> should have correct values', () => {
            expect(component.find('.dkreg-form-block__label').text()).toEqual(LanguageConstant.PASSWORD);
            expect(component.find('.dkreg-form-block__input').text()).toEqual('');
            expect(component.find('.dkreg-form-block__helper--error').text()).toEqual(`<IconExclamationCircle />${LanguageConstant.NAME_FIRST_ERROR}`);
        });
    });
})
