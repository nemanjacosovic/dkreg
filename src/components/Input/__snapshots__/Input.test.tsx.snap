// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="dkreg-form-block dkreg-form-block--input-name-first"
>
  <label
    className="dkreg-form-block__label"
    htmlFor="nameFirst"
  >
    Fornavn
  </label>
  <span
    className="dkreg-form-block__input-wrapper"
  >
    <input
      className="dkreg-form-block__input"
      id="nameFirst"
      name="nameFirst"
      placeholder="Harald"
      type="text"
    />
  </span>
  <div
    className="dkreg-form-block__helper"
  >
    <span
      className="dkreg-form-block__helper--error"
    >
      <IconExclamationCircle
        svgHeight="18"
      />
      Forkert v√¶rdi
    </span>
  </div>
</div>
`;
